tests/test_smart_bits.py

```diff
@@ -0,0 +1,79 @@
+import magma as m
+
+
+def test_circuit():
+
+    class _Foo(m.Circuit):
+        io = m.IO(
+            I0=m.In(m.SmartBits[7]),
+            I1=m.In(m.SmartBits[9]),
+            I2=m.In(m.SmartBits[12]),
+            O=m.Out(m.SmartBits[10]),
+            O2=m.Out(m.SmartBits[7]),
+            O3=m.Out(m.SmartBit),
+        )
+        x = io.I0 + io.I1 + io.I2
+        y = (io.I1 <= io.I2) + io.I0
+
+        print ()
+        print ("=======================")
+        print (x)
+        print (y)
+        print (io.I0)
+        print ("=======================")
+
+        io.O @= x
+        io.O2 @= y
+        io.O3 @= io.I0
+
+    print (repr(_Foo))
+
+    m.compile("Foo", _Foo, output="coreir-verilog", inline=True)
```

Comment:
It could imagine it being useful to have "pretty print" (maybe an extension of default circuit repr) that annotates the final widths that were determined

Reply:
(which would also be useful for testing)

Reply:
Makes sense. Will propose something in a upcoming commit.

Reply:
Done, attached `_smart_expr_` to the smart bits that get wired to. See `test_complex`.